CMakeLists.txt

```diff
@@ -5,125 +5,155 @@ IF(NOT CMAKE_BUILD_TYPE)
+set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
+set(CMAKE_CXX_EXTENSIONS OFF)
+cmake_policy(SET CMP0057 NEW) # enable IN_LIST operator
+if(cxx_std_11 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+  add_definitions(-DCOMPILEDWITHC11)
 endif()
 
+# Compiler specific options for gcc
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS}  -Wall   -O3")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall   -O3")
+  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -march=native")
+  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -march=native")
+  # set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wno-deprecated -O3 -march=native ")
+  # set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wno-deprecated -O3 -march=native")
+endif()
+# Compiler specific options for MSVC
+if (CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W3")
+  # use the static CRT if other libraries are compiled with /MT
+  # (CMake sets /MD by default, but for example Pangolin uses /MT by default)
+  #set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /MT")
+  #set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /MTd")
+endif()
+
+
 LIST(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake_modules)
 
-find_package(OpenCV 3)
+find_package(OpenCV 3 QUIET)
 if(NOT OpenCV_FOUND)
-   find_package(OpenCV 2.4.3 QUIET)
-   if(NOT OpenCV_FOUND)
+  find_package(OpenCV 4 QUIET)
+  if(NOT OpenCV_FOUND)
+    find_package(OpenCV 2.4.3 QUIET)
+    if(NOT OpenCV_FOUND)
       message(FATAL_ERROR "OpenCV > 2.4.3 not found.")
-   endif()
+    endif()
+  endif()
 endif()
-
-MESSAGE("OPENCV VERSION:")
-MESSAGE(${OpenCV_VERSION})
+MESSAGE(STATUS "OpenCV VERSION: ${OpenCV_VERSION}")
 
 find_package(Eigen3 3.1.0 REQUIRED)
+
 find_package(Pangolin REQUIRED)
 
+find_package(Boost REQUIRED COMPONENTS filesystem system serialization)
+MESSAGE(STATUS "Boost_LIBRARIES: ${Boost_LIBRARIES}")
+
+set(OPENSSL_USE_STATIC_LIBS TRUE)
+find_package(OpenSSL REQUIRED) # for crypto library
+
+
 include_directories(
-${PROJECT_SOURCE_DIR}
-${PROJECT_SOURCE_DIR}/include
-${PROJECT_SOURCE_DIR}/include/CameraModels
-${EIGEN3_INCLUDE_DIR}
-${Pangolin_INCLUDE_DIRS}
+  ${PROJECT_SOURCE_DIR}
+  ${PROJECT_SOURCE_DIR}/include
+  ${PROJECT_SOURCE_DIR}/include/CameraModels
+  ${EIGEN3_INCLUDE_DIR}
+  ${Pangolin_INCLUDE_DIRS}
+  ${OPENSSL_INCLUDE_DIR}
 )
 
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/lib)
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/lib)
+#set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/bin)
+
+set(ORB_SLAM3_LIB_TYPE SHARED)
+IF(WIN32)
+  # uncomment only one of the lines below:
+  # because there is no DLL API in the code,
+  # we either have to build STATIC libraries, 
+  set(ORB_SLAM3_LIB_TYPE STATIC)
+  # or we have to export all symbols from all SHARED libraries
+  #set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
+  # Note that the ideal solution would be to add an export API to the relevant classes.
```

Comment:
I would do the DLL interface in a separate merge request later because I also expect a lot of problems with the outdated g2o library. We should upgrade to a new version of g2o first, which already comes with a DLL interface

Reply:
About the first point, we should ask the original authors (@richard-elvira, @ccamposm) what is the intention here. It seems the key must be constant in the map, but I also think the Keyframe should not change, so maybe `const KeyFrame *const` is the correct entry here.

I correct the second point, thanks for pointing it out! I change it to std::size_t

Reply:
btw, I have Eigen 3.3.7, MSVC 16.7, and it has no problem with the original `const KeyFrame*`

another question to this: should be the key type in the first line (50) be the same as the key type in the second line (51)?